presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -15,6 +15,7 @@
 
 import com.facebook.presto.Session;
 import com.facebook.presto.connector.system.GlobalSystemConnector;
+import com.facebook.presto.execution.QueryManagerConfig.MaterializeExchangesStrategy;
```

Comment:
The materialization choice doesn't depend on any of the complicated logic in add exchanges.  Instead of adding still more complexity to AddExchanges, could we have a simple optimizer that runs after add exchanges and converts REMOTE exchanges to REMOTE_MATERIALIZED exchanges.  That will also make it simpler to introduce costing to that choice later.

Reply:
This is a very good comment.

We had a similar discussion with @wenleix. Eventually we decided that it is not a good idea to split the AddExchanges logic in multiple places. The AddExchanges optimizer is very complex, and we don't want to endup with two different classes that partially duplicate the same logic.

Although currently the logic is straightforward, it might not be so once we support materialized exchanges for the cases when one of the join tables is already bucketed.

Reply:
Based on offline discussions, it seems like the bigger issue is that ability to materialize depends on using connector partitioning, which has to be set in AddExchanges so that nodes higher up in the tree can derive the correct partitioning properties.  For now we'll keep the materialization decision here, and we'll extract it when we need to.